server/chat-plugins/rock-paper-scissors.ts

```diff
@@ -0,0 +1,404 @@
+		}
+		this.playerTable = {};
+	}
+	choose(user: User, option: string) {
+		const player = this.getPlayer(user);
+		if (player.currentChoice) throw new Chat.ErrorMessage("You have already made your choice!");
+		player.currentChoice = option;
+		this.add(Utils.html`${user.name} has made their choice.`);
+		this.resetOptions(user);
+		if (this.players.filter(item => item.currentChoice).length > 1) {
+			clearTimeout(this.roundTimer!);
+			return this.runMatch();
+		}
+		return true;
+	}
+	leaveGame(user: User) {
+		const player = this.getPlayer(user);
+		player.sendRoom(`You left the game.`);
+		delete this.playerTable[user.id];
+		this.end();
+	}
+	addPlayer(user: User) {
+		if (this.playerTable[user.id]) throw new Chat.ErrorMessage(`You are already a player in this game.`);
+		this.playerTable[user.id] = new RPSPlayer(user, this);
+		this.room.auth.set(user.id, Users.PLAYER_SYMBOL);
+		user.sendTo(this.room, `You have successfully joined the Rock, Paper, Scissors game.`);
+		return this.playerTable[user.id];
+	}
+}
+
+function getGame(context: CommandContext) {
+	const room = context.requireRoom();
+	const game = room.getGame(RPSGame);
+	if (!game) throw new Chat.ErrorMessage(`There is no Rock, Paper, Scissors game going on in this room.`);
+	if (game.gameid !== 'rockpaperscissors') {
+		throw new Chat.ErrorMessage(`A game of ${game.title} is already going on.`);
+	}
+	return game;
+}
+
+function findExisting(user1: string, user2: string) {
+	return Rooms.get(`rps-${user1}-${user2}`) || Rooms.get(`rps-${user2}-${user1}`);
+}
+
+export const commands: ChatCommands = {
+	rps: 'rockpaperscissors',
+	rockpaperscissors: {
+		challenge: 'create',
+		create(target, room, user) {
+			target = target.trim();
+			const [rounds, userid] = target.split(',');
+			if (!rounds || !target) {
+				return this.parse('/help rps');
+			}
+			const roundsNum = parseInt(target);
+			if (isNaN(roundsNum)) {
+				return this.errorReply(`Invalid round number.`);
+			}
+			if (roundsNum > MAX_ROUNDS) {
+				return this.errorReply(`You cannot set more than ${MAX_ROUNDS} rounds.`);
+			}
+			const targetUser = userid ? Users.get(userid) : this.pmTarget;
+			if (targetUser === user) return this.errorReply(`You cannot challenge yourself.`);
+			if (!targetUser) {
+				return this.errorReply(
+					`User ${this.targetUsername} not found. Either specify a username or use this command in PMs.`
+				);
+			}
+			const options = {
+				modchat: '+',
+				isPrivate: true,
+			};
+			const existingRoom = findExisting(user.id, targetUser.id);
+			const roomid = existingRoom ? existingRoom.roomid : `rps-${user.id}-${targetUser.id}`;
+			if (existingRoom?.game && !existingRoom.game.ended) {
+				return this.errorReply(`You already have a Rock, Paper, Scissors game against ${targetUser.name}.`);
+			}
+			const gameRoom = existingRoom ? existingRoom : Rooms.createGameRoom(
+				roomid as RoomID, `[RPS] ${user.name} vs ${targetUser.name}`, options
+			);
```

Comment:
The game room should probably be created when the challenge is accepted, not when it's made, like how battles work.